flocker/node/agents/cinder.py

```diff
@@ -227,46 +227,141 @@ def list():
+        self.expected_volume = expected_volume
+        self.desired_state = desired_state
+        self.elapsed_time = elapsed_time
+
+    def __str__(self):
+        return (
+            'Timed out while waiting for volume. '
+            'Expected Volume: {!r}, '
+            'Expected State: {!r}, '
+            'Elapsed Time: {!r}'.format(
+                self.expected_volume, self.desired_state, self.elapsed_time)
+            )
 
-    :param ICinderVolumeManager volume_manager: An API for listing volumes.
-    :param Volume expected_volume: The ``Volume`` to wait for.
-    :param unicode expected_status: The ``Volume.status`` to wait for.
-    :param int time_limit: The maximum time, in seconds, to wait for the
-        ``expected_volume`` to have ``expected_status``.
-    :raises Exception: If ``expected_volume`` with ``expected_status`` is not
-        listed within ``time_limit``.
+
+class UnexpectedStateException(Exception):
+    """
+    An unexpected state was encountered by a volume as a result of operation.
+    """
+    def __init__(self, expected_volume, desired_state, unexpected_state):
+        self.expected_volume = expected_volume
+        self.desired_state = desired_state
+        self.unexpected_state = unexpected_state
+
+    def __str__(self):
+        return (
+            'Unexpected state while waiting for volume. '
+            'Expected Volume: {!r}, '
+            'Expected State: {!r}, '
+            'Reached State: {!r}'.format(
+                self.expected_volume, self.desired_state,
+                self.unexpected_state)
+            )
+
+
+class VolumeStateMonitor:
+    """
+    Monitor a volume until it reaches a nominated state.
+    :ivar ICinderVolumeManager volume_manager: An API for listing volumes.
+    :ivar Volume expected_volume: The ``Volume`` to wait for.
+    :ivar unicode desired_state: The ``Volume.status`` to wait for.
+    :ivar transient_states: A sequence of valid intermediate states.
+        The states must be listed in the order that are expected to occur.
+    :ivar int time_limit: The maximum time, in seconds, to wait for the
+        ``expected_volume`` to have ``desired_state``.
+    :raises: UnexpectedStateException: If ``expected_volume`` enters an
+        invalid state.
+    :raises TimeoutException: If ``expected_volume`` with
+        ``desired_state`` is not listed within ``time_limit``.
     :returns: The listed ``Volume`` that matches ``expected_volume``.
     """
-    start_time = time.time()
-    # Log stuff happening in this loop.  FLOC-1833.
-    while True:
+    def __init__(self, volume_manager, expected_volume,
+                 desired_state, transient_states=(), time_limit=60):
+        self.volume_manager = volume_manager
+        self.expected_volume = expected_volume
+        self.desired_state = desired_state
+        self.transient_states = transient_states
+        self.time_limit = time_limit
+        self.start_time = time.time()
```

Comment:
I looked at this, but there is no reason that the `wait_for_volume_state` would pass the current time in general. It only makes sense when the code knows a timeout is possible.  I think it would make more sense to pass a Clock object to the VolumeStateMonitor, which knows about timeouts.  Although this might help in future testing, it doesn't seem necessary right now, especially as there are several issues (FLOC-1831,-1832,-1833) which may change this code significantly later. (I'd like the EC2 and Cinder code to share more at some point).